src/cfg/Exceptions.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 1996-2020 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID__SRC_CFG_EXCEPTIONS_H
+#define SQUID__SRC_CFG_EXCEPTIONS_H
+
+#include "sbuf/Stream.h"
+
+#include <limits>
+#include <stdexcept>
+
+namespace Cfg
+{
+
+/**
+ * Helper class to distinguish serious parsing errors which need
+ * Administrative attention before Squid can operate.
+ */
+class FatalError : public std::exception
```

Comment:
TextException requires and embeds code files and line numbers into the output which is both irrelevant and interferes with interpretation of these config parser exceptions. I saw no need to combine the two and potentially allow catch statements for TextException to handle these config errors through implicit casting.

Reply:
> files and line numbers [are] both irrelevant and interferes with interpretation of these config parser exceptions

I disagree on both counts:

1. Source code location is often _very_ useful in interpreting a configuration error, especially because many of our errors do not have enough information to understand what exactly went wrong and locating the relevant parsing code is often difficult, even for seasoned developers.
2. Source code location does not interfere with interpretation of an exception -- it is just additional information. The code will ignore it. Humans may ignore it (it is their call).

> I saw no need to combine the two

The reason is that both classes support the same basic need -- report of a text-based error condition with a potential of adding (nested) scope information as a likely future enhancement. We should not duplicate/fork that support.

> and potentially allow catch statements for TextException to handle these config errors through implicit casting.

The catching code can still distinguish different kinds of errors, of course. Catch-clause matching will continue to work if we needed it. In fact, both current FatalError and TextException are (direct or indirect) children of the same std::exception class so this principle already works in this PR.

If there is some specific roadblock that makes TextException reuse (direct or in a child class) difficult, let's discuss. I may be able to help clear that roadblock.

Reply:
When I am done testing the overflow protection changes I will paste logs of the parser output here so you can see the UI pollution from TextException.

Reply:
Sounds good. If we agree that what you paste is indeed pollution, then I am sure we can find a way to address it without forking TextException.

Reply:
Trivial broken squid.conf has:
```
http_port 99999

#
#
https_port 88888
```

This is the output from master right now:

* k check:
```
(time)| FATAL: Bungled ./msg_test.conf line 2: http_port 99999
(time)| Squid Cache (Version 6.0.0-VCS): Terminated abnormally.
```

* k parse:
```
(time)| Processing Configuration File: ./msg_test.conf (depth 0)
(time)| Processing: http_port 99999
(time)| ERROR: The value '99999' is larger than the type 'short'.
(time)| Not currently OK to rewrite swap log.
(time)| storeDirWriteCleanLogs: Operation aborted.
(time)| FATAL: Bungled ./msg_test.conf line 2: http_port 99999
(time)| Squid Cache (Version 6.0.0-VCS): Terminated abnormally.
```

  NP: we have had several reports from confused users who noticed the FATAL, but missed the ERROR message - which is sometimes a dozen or more lines earlier in the log as components displayed their cleanup.
  This does not directly counter your point, but does provide a first hint that admin are not good at ignoring irrelevant details around the logged errors.


Output from this PR when throwing a "class FatalError : public TextException":

* k check
```
(time)| invalid value: 99999 must be below 65535
    exception location: /cfg/Exceptions.h(28) FatalError
(time)| FATAL: Bungled ./msg_test.conf line 2: http_port 99999
(time)| Squid Cache (Version 6.0.0-VCS): Terminated abnormally.
```

* k parse
```
(time)| Processing Configuration File: ./msg_test.conf (depth 0)
(time)| Processing: http_port 99999
(time)| invalid value: 99999 must be below 65535
    exception location: /cfg/Exceptions.h(28) FatalError
(time)| Processing: https_port 88888
(time)| invalid value: 88888 must be below 65535
    exception location: /cfg/Exceptions.h(28) FatalError
```

 NP: I also tried the idea of having Cfg::FatalError be an inline function that throws a TextException. It produces identical output to the above - including the garbage about cfg/Exceptions.h.


Output from above but with FatalError adding the documented labels to the output:

* k check:
```
(time)| FATAL: invalid value: 99999 must be below 65535
    exception location: /cfg/Exceptions.h(42) FatalError
(time)| FATAL: Bungled ./msg_test.conf line 2: http_port 99999
(time)| Squid Cache (Version 6.0.0-VCS): Terminated abnormally.
```

* k parse:
```
(time)| Processing Configuration File: ./msg_test.conf (depth 0)
(time)| Processing: http_port 99999
(time)| FATAL: invalid value: 99999 must be below 65535
    exception location: /cfg/Exceptions.h(42) FatalError
(time)| Processing: https_port 88888
(time)| FATAL: invalid value: 88888 must be below 65535
    exception location: /cfg/Exceptions.h(42) FatalError
```


Output from this PR currently:

* k check:
```
(time)| FATAL: invalid value: 99999 must be below 65535
(time)| FATAL: Bungled ./msg_test.conf line 2: http_port 99999
(time)| Squid Cache (Version 6.0.0-VCS): Terminated abnormally.
```

* k parse
```
(time)| Processing Configuration File: ./msg_test.conf (depth 0)
(time)| Processing: http_port 99999
(time)| FATAL: invalid value: 99999 must be below 65535
(time)| Processing: https_port 88888
(time)| FATAL: invalid value: 88888 must be below 65535
```

Reply:
I like this!

Reply:
> This does not directly counter your point, but does provide a first hint that admin are not good at ignoring irrelevant details around the logged errors.

Fortunately, nobody has suggested logging irrelevant details. I am sure that we can polish the output from parsing errors so that it provides all the relevant details that do not distract a reasonable admin. The current output (_all_ the versions you have pasted) does not do that IMO.

Please also note that insufficiently detailed output is just one part of the problem here. It is wrong to fork TextException for architectural reasons as well.

Please let me know if I should work on adjusting FatalError to address my concerns.

Reply:
> > This does not directly counter your point, but does provide a first hint that admin are not good at ignoring irrelevant details around the logged errors.
> 
> Fortunately, nobody has suggested logging irrelevant details. 

Then I have misunderstood. Are you not requesting that I use TextException for these exceptions?

> I am sure that we can polish the output from parsing errors so that it provides all the relevant details that do not distract a reasonable admin. The current output (_all_ the versions you have pasted) does not do that IMO.
> 

"exception location: /cfg/Exceptions.h(42) FatalError" is garbage. It never changes. *all* exceptions contain that exact string. It has no use and _doubles_ the size of the output.

> Please also note that insufficiently detailed output is just one part of the problem here. It is wrong to fork TextException for architectural reasons as well.
> 

What are calling a "fork" in here?

> Please let me know if I should work on adjusting FatalError to address my concerns.

Insufficient detail is not the issue here. If you want more *relevant* details added we can discuss that on a per-message basis. Some might be done in this PR, the ones that require extra infrastructure should wait for later.

Reply:
> Are you not requesting that I use TextException for these exceptions?

I am requesting that FatalError and all similar new text-based exception classes, if any, are inherited from TextException. Needless to say, the TextException class itself may be adjusted or even refactored/split/etc. to properly accommodate new use cases.

> "exception location: /cfg/Exceptions.h(42) FatalError" is garbage. It never changes. all exceptions contain that exact string. It has no use and doubles the size of the output.

I think we are in agreement here! That agreement does not mean that the source code location where the configuration parsing problem was discovered is garbage. The `cfg/Exceptions.h` location in your example is indeed essentially garbage, but that is not what we should be reporting, of course. I was talking about the idea, not a bad implementation of it. Just because it is possible to misuse X, does not mean that X is useless.

> What are calling a "fork" in here?

I used that word to describe a phenomenon where the same basic concept (a text-based Squid-emitted exception) is copied from an existing class into a new independent class, allowing each to evolve independently (with all the advantages and problems that such independent existence brings). (It is very similar to repository forking.) The alternatives include inheritance and encapsulation.

> Insufficient detail is not the issue here.

It _is_ one of the issues for me.

> If you want more relevant details added we can discuss that on a per-message basis.

My primary concerns are not message-specific. They are architectural/global. Virtually all reports about configuration parsing errors should include the source code location where the error was detected because that information is often very useful and because the concept of source code location collection and reporting is already supported in Squid.